espnet/nets/pytorch_backend/e2e_asr_transformer.py

```diff
@@ -142,33 +166,83 @@ def forward(self, xs_pad, ilens, ys_pad):
         '''
-        # forward encoder
+        # 1. forward encoder
         xs_pad = xs_pad[:, :max(ilens)]  # for data parallel
         src_mask = (~make_pad_mask(ilens.tolist())).to(xs_pad.device).unsqueeze(-2)
         hs_pad, hs_mask = self.encoder(xs_pad, src_mask)
         self.hs_pad = hs_pad
 
-        # forward decoder
+        # 2. forward decoder
         ys_in_pad, ys_out_pad = self.add_sos_eos(ys_pad)
         ys_mask = self.target_mask(ys_in_pad)
         pred_pad, pred_mask = self.decoder(ys_in_pad, ys_mask, hs_pad, hs_mask)
         self.pred_pad = pred_pad
 
-        # compute loss
+        # 3. compute attenttion loss
         loss_att = self.criterion(pred_pad, ys_out_pad)
         self.acc = th_accuracy(pred_pad.view(-1, self.odim), ys_out_pad,
                                ignore_label=self.ignore_id)
 
-        # TODO(karita) show predected text
+        # TODO(karita) show predicted text
         # TODO(karita) calculate these stats
-        cer, cer_ctc, wer = 0.0, 0.0, 0.0
-        if self.ctc is None:
+        # 4. compute ctc loss   || compute cer without beam search
+        if self.mtlalpha == 0:
             loss_ctc = None
+            cer_ctc = None
         else:
             batch_size = xs_pad.size(0)
             hs_len = hs_mask.view(batch_size, -1).sum(1)
             loss_ctc = self.ctc(hs_pad.view(batch_size, -1, self.adim), hs_len, ys_pad)
 
+            cers = []
+
+            y_hats = self.ctc.argmax(hs_pad.view(batch_size, -1, self.adim)).data
+            for i, y in enumerate(y_hats):
+                y_hat = [x[0] for x in groupby(y)]
+                y_true = ys_pad[i]
+
+                seq_hat = [self.char_list[int(idx)] for idx in y_hat if int(idx) != -1]
+                seq_true = [self.char_list[int(idx)] for idx in y_true if int(idx) != -1]
+                seq_hat_text = "".join(seq_hat).replace(self.space, ' ')
+                seq_hat_text = seq_hat_text.replace(self.blank, '')
+                seq_true_text = "".join(seq_true).replace(self.space, ' ')
+
+                hyp_chars = seq_hat_text.replace(' ', '')
+                ref_chars = seq_true_text.replace(' ', '')
+                if len(ref_chars) > 0:
+                    cers.append(editdistance.eval(hyp_chars, ref_chars) / len(ref_chars))
+
+            cer_ctc = sum(cers) / len(cers) if cers else None
+
+        # 5. compute cer/wer
+        if self.training or not (self.report_cer or self.report_wer):
+            cer, wer = 0.0, 0.0
+            # oracle_cer, oracle_wer = 0.0, 0.0
+        else:
+            word_eds, word_ref_lens, char_eds, char_ref_lens = [], [], [], []
+            y_hats = pred_pad.argmax(dim=2)
+            for i, y_hat in enumerate(y_hats):
+                y_true = ys_pad[i]
+                eos_true = np.where(y_true.cpu() == -1)[0]
+                eos_true = eos_true[0] if len(eos_true) > 0 else len(y_true)
+                seq_hat = [self.char_list[int(idx)] for idx in y_hat[:eos_true]]
```

Comment:
Why did you truncate the predicted character sequence? This may underestimate the edit distance.

Reply:
Is because the output has not padded outs with `-1`, so when it is compared with the true, the WER is even higher than the one obtained from the decoding.

Reply:
I see. If you leave a comment about this on the source code, it is great.